### PR TITLE
Feature/profile

### PR DESCRIPTION
**Pull request by Gimleux**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -17,10 +17,12 @@
     "axios": "0.21.1",
     "bootstrap": "4.6.0",
     "eslint-plugin-react-hooks": "4.2.0",
+    "jquery": "^1.9.1",
     "node-sass": "4.14.1",
-    "react": "17.0.1",
+    "popper.js": "^1.16.1",
```
Comment (qvalentin) [inline package.json]: jquery and popper can be removed

File: src/background/api/userInformation.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+export interface UserInformation {
+    userId: number | null;
+    username?: string | null;
+    groups?: number[]|null;
+    password?: string;
+    confirmationPassword?: string;
+}
+
+export const changeUserInformation = (userWithNewInformation: UserInformation): Promise<UserState> => {
+    console.log("User given tu update user api:")
```
Comment (qvalentin) [inline src/background/api/userInformation.ts]: "Tu" doch nicht so was

File: src/background/api/userInformation.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    console.log(userWithNewInformation)
+    return new Promise((resolve, reject) => {
+        return Axios.put(`${hostname}${userPath}/${userWithNewInformation.userId}/edit`, userWithNewInformation)
+            .then((response: AxiosResponse<UserState>) => {
+                store.dispatch(updateUser(JSON.parse(response.config.data)));
```
Comment (qvalentin) [inline src/background/api/userInformation.ts:23]: Why do you parse the json?

File: src/background/redux/actions/user.tsx [hunk 4]
```diff
@@ -1,6 +1,11 @@
...
 
 export const addUser = (content: UserState):AddUser => ({
     type: ADD_USER,
     payload: content
 });
+
+export const updateUser = (content: UserState):UpdateUser => ({
+    type: UPDATE_USER,
+    payload: content
+})
```
Comment (qvalentin) [inline src/background/redux/actions/user.tsx]: One of them is not necessary as they do the same thing, so remove one,

File: src/background/redux/reducers/user.tsx [hunk 5]
```diff
@@ -1,18 +1,23 @@
...
-      console.log("[Redux] adding user");
-      return action.payload;
+    switch (action.type) {
+        case ADD_USER: {
+            console.log("[Redux] adding user");
+            return action.payload;
+        }
+        case UPDATE_USER: {
+            console.log("[Redux] updating user");
+            return action.payload;
```
Comment (qvalentin) [inline src/background/redux/reducers/user.tsx]: Same here, why do a switch, when both functions do the same.

File: src/components/pages/User/Profile.tsx [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    const [alertVisibility, setAlertVisibility] = useState<boolean>(false);
+
+    const handleAlertVisibility = (duration: number, color: "primary" | "secondary" | "success" | "danger" | "warning" | "info" | "light" | "dark", message: string) => {
+        if (!alertVisibility) {
+            setAlertMessage(message);
+            setAlertColor(color);
+            setAlertVisibility(true);
+            setTimeout(() => {
+                setAlertVisibility(false);
+            }, duration);
```
Comment (qvalentin) [inline src/components/pages/User/Profile.tsx:26]: nice solution

File: src/components/pages/User/Profile.tsx [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+            //     handleAlertVisibility(DEFAULT_ALERT_DURATION, "danger", "Error: Please pay attention to the notes below the input fields.")
+            return;
+        }
+        newUser["username"] = inputUser.username;
+        if (inputUser.password || inputUser.passwordConfirmation) {
+            if (inputUser.password !== inputUser.passwordConfirmation) {
+                handleAlertVisibility(DEFAULT_ALERT_DURATION, "danger", "Error: Password and password confirmation must match.")
+                return;
+            }
+            if (inputUser.password.match(/\d/) == null || inputUser.password.match(/[a-z]/) == null || inputUser.password.match(/[A-Z]/) == null || notMinStrLength(inputUser.password, MIN_PASSWORD_LENGTH)) {
```
Comment (qvalentin) [inline src/components/pages/User/Profile.tsx:55]: can this please be a function that we can use also for registration?

File: src/components/pages/User/Profile.tsx [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+            newUser["confirmationPassword"] = inputUser.passwordConfirmation
+        }
+
+        await changeUserInformation(newUser)
+            .then(res => {
+                changeEditMode()
+                handleAlertVisibility(DEFAULT_ALERT_DURATION, "success", "Worked: " + (res));
+            })
+            .catch(err => {
+                handleAlertVisibility(DEFAULT_ALERT_DURATION, "danger", "Error: " + (err.outputMessage ? err.outputMessage : (err.httpStatus + " " + err.httpMessage)))
```
Comment (qvalentin) [inline src/components/pages/User/Profile.tsx]: image

File: src/components/pages/User/Profile.tsx [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+                    })}
+                    </dd>
+                </dl>
+            </Container>
+        )
+    }
+
+    return (
+        <Container className="page-content">
+            <Container fluid className=" mt-1 mb-3">
```
Comment (qvalentin) [inline src/components/pages/User/Profile.tsx]: more nested Containers

File: src/components/pages/User/Registration.tsx [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+export default function Registration(): ReactElement {
+    const [alertMessage, setAlertMessage] = useState<string>("Error 404: No Message found.");
+    const [alertVariant, setAlertColor] = useState<"primary" | "secondary" | "success" | "danger" | "warning" | "info" | "light" | "dark">("success");
+    const [alertVisibility, setAlertVisibility] = useState<boolean>(false);
+
+    // const [newUser, setNewUser] = useState<UserInformationInterface>({
+    //     username: "",
+    //     password: "",
+    //     passwordConfirmation: ""
+    // })
```
Comment (qvalentin) [inline src/components/pages/User/Registration.tsx]: clean this up

File: src/components/pages/User/Registration.tsx [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+                const leftContainerOffset: number = container?.getBoundingClientRect().left ?? 0;
+
+                let containerPadding: string | number | null = getStyleValue(container, "padding-left");
+                const pxPosition = containerPadding.indexOf("px");
+                containerPadding = pxPosition === -1 ? null : Number(containerPadding.substr(0, pxPosition))
+
+                logo.style.left = -(leftContainerOffset + logo.offsetWidth * 2 + (containerPadding ?? 20)) + "px";
+            }
+        }
+
```
Comment (qvalentin) [inline src/components/pages/User/Registration.tsx:33]: wtf is this

File: src/components/pages/User/Registration.tsx [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+    const handleSubmit = async (newUser:UserInformationInputInterface) => {
+        console.log("[REGISTRATION] handleSubmit")
+        if (!newUser.username) {
+            handleAlertVisibility(DEFAULT_ALERT_DURATION, "danger", "Error: Please choose an username.")
+        } else if (newUser.password !== newUser.passwordConfirmation) {
+            handleAlertVisibility(DEFAULT_ALERT_DURATION, "danger", "Error: Password and password confirmation must match.")
+        } else if (newUser.password.match(/\d/) == null || newUser.password.match(/[a-z]/) == null || newUser.password.match(/[A-Z]/) == null || notMinStrLength(newUser.password, MIN_PASSWORD_LENGTH)) {
+            handleAlertVisibility(DEFAULT_ALERT_DURATION, "danger", "Error: Please pay attention to the notes below the input fields.")
+        } else {
+            await registerNewUser(newUser.username, newUser.password, newUser.passwordConfirmation)
```
Comment (qvalentin) [inline src/components/pages/User/Registration.tsx]: is await even needed here?

File: src/components/pages/User/Registration.tsx [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+            setTimeout(() => {
+                setAlertVisibility(false);
+            }, duration);
+        }
+    }
+
+    return (
+        <Container className="h-100" style={{position: "relative"}} id="registrationContainer">
+            <Row>
+                <Col md={{span: 6, offset: 3}}>
```
Comment (qvalentin) [inline src/components/pages/User/Registration.tsx]: image
looks weird

File: src/components/pages/User/Profile.tsx [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+
+    function DisplayProfile(): ReactElement {
+        return (
+            <Container className="profile-information-display p-0">
+                <h2 className="h3 pb-3">
+                    {user.username}
+                </h2>
+                <dl>
+                    <dt>Username</dt>
+                    <dd>{user.username}</dd>
```
Comment (qvalentin) [inline src/components/pages/User/Profile.tsx:98]: who uses dl and dt XDDDDDDDDD

File: src/components/pages/User/UserInformationInput.tsx [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+        let value: [string, boolean] | string = makePasswordInputFitRules(event.target.value);
+        if (!value[1]) {
+            value = password;
+        } else {
+            value = value[0]
+        }
+        setPasswordInformationLength(!notMinStrLength(value, MIN_PASSWORD_LENGTH));
+        setPasswordInformationLowercase(value.match(/[a-z]/) !== null);
+        setPasswordInformationUppercase(value.match(/[A-Z]/) !== null);
+        setPasswordInformationNumber(value.match(/\d/) !== null);
```
Comment (qvalentin) [inline src/components/pages/User/UserInformationInput.tsx]: redundant, this code should only be defined once

File: src/components/pages/User/UserInformationInput.tsx [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+        setPassword(value)
+    }
+
+    const handlePasswordConfirmationChange = async (event: ChangeEvent<HTMLInputElement>) => {
+        event.preventDefault();
+        let value: [string, boolean] | string = makePasswordInputFitRules(event.target.value);
+        if (!value[1]) {
+            value = passwordConfirmation;
+        } else {
+            value = value[0]
```
Comment (qvalentin) [inline src/components/pages/User/UserInformationInput.tsx]: why such a complicated return value? Could be simpler

**Review by qvalentin — CHANGES_REQUESTED** [review]

File: src/components/pages/User/Registration.tsx (hunk 11 above)
Comment (Gimleux) [reply src/components/pages/User/Registration.tsx:33]: funny how this code was not even changed by this pr (only repositioned)

File: src/background/api/userInformation.ts (hunk 3 above)
Comment (Gimleux) [reply src/background/api/userInformation.ts:23]: the changed user is part of the api response in json file format

File: src/components/pages/User/Registration.tsx (hunk 11 above)
Comment (Gimleux) [reply src/components/pages/User/Registration.tsx:33]: but btw, it is the function to set the first position (depending on the screen width) of the moving image which is played when a new user is registered

File: src/components/pages/User/UserInformationInput.tsx (hunk 15 above)
Comment (Gimleux) [reply src/components/pages/User/UserInformationInput.tsx]: it is connected to some css, will talk about this next time

**Review by qvalentin — APPROVED** [review]

**Pull request merged**